tokio-io/src/split.rs

```diff
@@ -12,12 +13,44 @@ pub struct ReadHalf<T> {
     handle: BiLock<T>,
 }
 
+impl<T: AsyncRead + AsyncWrite> ReadHalf<T> {
+    /// Reunite with a previously split `WriteHalf`.
+    ///
+    /// If this `ReadHalf` and the given `WriteHalf` do not originate from
+    /// the same `AsyncRead::split` operation they will be returned, wrapped
+    /// in an `UnsplitError`.
+    pub fn unsplit(self, w: WriteHalf<T>) -> Result<T, UnsplitError<T>> {
```

Comment:
I know that `BiLock` returns `Result`, but I'm wondering if it would be better to panic on error here. The error case is always due to a bug vs. a runtime error case.

Reply:
I would prefer not to panic and not to reason about internal error conditions of other libraries which may change in future versions. `BiLock::reunite` returns a `Result` and therefore I think we should not make assumptions about its inner workings but just return a `Result` as well.

Reply:
How would you handle the `Err` case in a way that makes sense?

Reply:
What do you mean by "makes sense"? If you do not like `UnsplitError`, what about the original version which returned `Result<T, (ReadHalf<T>, WriteHalf<T>)>`?

Reply:
My general philosophy is to panic on errors that result from code bugs. If two unrelated handles are passed to `unsplit`, then assumptions elsewhere in the code are incorrect and there isn't much the code can do to recover from the bug (as it does not know what assumptions are incorrect). Tearing down state is the only "safe" option.

Reply:
I agree with that general philosophy but here we are calling the function of some other crate which gives us a `Result` and while code inspection convinces us that the error case can only occur when applying the function to two unrelated handles, this inspection only covers the current version of that crate. It may be unlikely that `BiLock::reunite` changes in this regard but I generally prefer to not rely on this.

Reply:
Looking more closely at `BiLock`'s `ReuniteError` it is described as:

> Error indicating two `BiLock<T>`s were not two halves of a whole, and thus could not be `reunite`d.

Since this can only happen due to a logic error I am ready to panic here ;-)